src/app/qgsmaptoolreshape.cpp

```diff
@@ -100,6 +100,54 @@ void QgsMapToolReshape::cadCanvasReleaseEvent( QgsMapMouseEvent * e )
         reshapeReturn = geom->reshapeGeometry( points() );
         if ( reshapeReturn == 0 )
         {
+          //avoid intersections on polygon layers
+          if (vlayer->geometryType() == QGis::Polygon) {
+
+            //ignore all current layer features as they should be reshaped too
+            QMap<QgsVectorLayer*, QSet<QgsFeatureId>> ignoreFeatures;
+            ignoreFeatures.insert(vlayer, vlayer->allFeatureIds());
+
+            int avoidIntersectionsReturn = geom->avoidIntersections(ignoreFeatures);
+            if ( avoidIntersectionsReturn == 1 )
+            {
+              //not a polygon type. Impossible to get there
+            }
+#if 0
```

Comment:
Thans for reviews @m-kuhn 

Yes this was originally copied from here : https://github.com/qgis/QGIS/blob/576875e9982270639d2ec450cfceb4e80f476dbe/src/app/qgsmaptooladdfeature.cpp#L256

And adapted for my needs. Note that : 
* Here, we need to call `vlayer->destroyEditCommand();`
* Message are lightly differents (cannot be added => cannot be reshaped).
* Messages should be emit from the current class.
So it would be complicated to factorize this "error handling".

This code does not seems to be up to date, QgsGeometry->avoidIntersections() only return 0 or 1:
https://github.com/qgis/QGIS/blob/14c858a0b1b56586902daaddb54840ac417f975c/src/core/geometry/qgsgeometry.cpp#L1499:L1514
I propose to remove parts of code related to avoidIntersectionsReturn == [2|3] from my pull request.
And possibly propose another for qgsmaptooladdfeature.

Regards